Using property CustomSlackChannel from logEvent as channel name override

seirysjs opened the pull request:
If such property exists in `logEvent` properties, then use it as slack channel name instead.
This way each logEvent message can have it's own channel specified.

NerijusD commented:
@mgibas Hello, that is needed feature. Who could review and merge this?

seirysjs commented:
This also allows scoped Logs to override channel name
for example:
```
using (_logger.BeginScope("{CustomSlackChannel}", customSlackChannel))
{
    _logger.LogCritical(exceptionToLog, message, args);
}
```

seirysjs commented:
Please review this, it's highly required feature 
@augustoproiete @bartelink @khellang

bartelink commented on Serilog.Sinks.Slack/SlackSink.cs:
```diff
@@ -115,7 +115,7 @@ protected Message CreateMessage(LogEvent logEvent)
                 {
                     Text = textWriter.ToString(),
-                    Channel = _options.CustomChannel,
+                    Channel = logEvent.Properties.ContainsKey("CustomSlackChannel") ? logEvent.Properties["CustomSlackChannel"].ToString().Replace("\"", string.Empty) : _options.CustomChannel,
```
TryGetValue and a pattern match?

seirysjs replied:
@bartelink changed to use `TryGetValue` with `pattern matching`. Would this be acceptable? if not, please could you provide an example what you have in mind exactly?

bartelink replied:
Looks better to me now 👍 

NOTE I'm not a repo owner/maintainer and don't have much insight into the ground rules/usage of the sink, so unfortunately I'm no use to you in getting to your overall goal...

seirysjs replied:
thanks for your input in reviewing this

seirysjs replied:
@bartelink could you please tell me who are the maintainers of this repo? and who is owner? so i could tag them to merge this

bartelink replied:
No idea; in general I'd avoid atting people and let them get notified; people have day jobs...

seirysjs replied:
Well, thanks for your time. I apologize for disturbing your working hours

bartelink replied:
I personally don't happen to mind atm (I'm doing a mishmash of stuff including lots of public github stuff today).

But for anyone who created a sink, brought it to the world, and maybe moved to a company with a different log rig, seeing a message in your inbox from an @ 20 mins after an issue is created might provoke a "Oh noes, I owe the internet stuff!"

seirysjs replied:
I will keep that in mind 👍

TrapperHell commented:
Hi @seirysjs, thank you for your contribution.

As bartelink pointed out, please allow a reasonable amount of time before pinging to get something done. We all have jobs and work to do.

While using this library myself I have had a use-case to get the logger to write to different channels - and I got it done without resorting to this. 

I will consider your approach and try to balance the easiness which it adds versus any potential pitfalls / feature-bloating. I will keep you posted.

TrapperHell commented on README.md:
````diff
@@ -63,3 +63,11 @@ Log.Logger = new LoggerConfiguration()
     .CreateLogger();
 ```
+
+Scope overriden channel name:
````
Typo in `overriden`. Should be `overridden`.

seirysjs replied:
typo fixed

NerijusD commented:
> Hi @seirysjs, thank you for your contribution.
> 
> As bartelink pointed out, please allow a reasonable amount of time before pinging to get something done. We all have jobs and work to do.
> 
> While using this library myself I have had a use-case to get the logger to write to different channels - and I got it done without resorting to this.
> 
> I will consider your approach and try to balance the easiness which it adds versus any potential pitfalls / feature-bloating. I will keep you posted.

@TrapperHell Could you share your approach, how did you do it? Did you have predefined channels? In our use case we want to have dynamic channel names on each run for our application.

seirysjs commented:
> Hi @seirysjs, thank you for your contribution.
> 
> As bartelink pointed out, please allow a reasonable amount of time before pinging to get something done. We all have jobs and work to do.
> 
> While using this library myself I have had a use-case to get the logger to write to different channels - and I got it done without resorting to this.
> 
> I will consider your approach and try to balance the easiness which it adds versus any potential pitfalls / feature-bloating. I will keep you posted.

Thank you for your review and insight. I will do my best to not tag people like this in the future anymore. I'm sorry for the disturbance in regards to this PR.

Please do share your approach to have the logger to write to different channels. I would really appreciate it!

TrapperHell commented:
As requested, here is the approach that I have used to cater for my need. However please note that this does not support dynamic, runtime slack channel changes.

> Did you have predefined channels? In our use case we want to have dynamic channel names on each run for our application.

Yes, the channels were predefined.

The approach is as follows, and powered by the [Serilog Expressions](https://github.com/serilog/serilog-expressions) library:

Define a logger per channel that you intend to log to, you can do this in the JSON configuration itself, such as:

```json
{
  "Name": "Logger",
  "Args": {
    "configureLogger": {
      "Filter": [
        {
          "Name": "ByIncludingOnly",
          "Args": {
            "expression": "EventId.Id = 1"
          }
        }
      ],
      "WriteTo": [
        {
          "Name": "Slack",
          "Args": { "Configure Slack sink configuration here..." }
        }
      ]
    }
  }
},
{
  ... More loggers here...
}
```

and then to log you would simply use:

```csharp
logger.LogInformation(new EventId(1), "Greetings...");
```

In reality I used another function to resolve the EventId necessary for which message, since it wasn't really dependent on log severity but business-logic.

I will review your changes during this weekend and let you know. Overall it looks fine, however I'm hoping there's a somewhat better options than using `.Replace("\"", string.Empty)` , and I might be a little pedantic here, but perhaps call the parameter `{ChannelOverride}` instead of `{CustomSlackChannel}`

seirysjs commented:
PR updated with solution to `ShowPropertyAttachments` exposing properties that are ment to be used as config overrides.
Added overrides for `CustomUserName` and `CustomIcon` as well. Renamed the property  `CustomSlackChannel` to `CustomChannel`. 
Reusing `SlackSinkOptions` property names that are meant to be overridden.
To override the config properties, it's now necessary to specify them in `PropertyOverrideList` that was newly added to `SlackSinkOptions`.

NerijusD commented on Serilog.Sinks.Slack/Models/SlackSinkOptions.cs:
```diff
@@ -94,6 +95,11 @@ public class SlackSinkOptions
+        /// <summary>
+        /// Optional: A list of <see cref="SlackSinkOptions"> properties that are being overridden by properties in the messages.
+        /// </summary>
+        public List<OverridableProperties> PropertyOverrideList { get; set; }
```
Could this be `HashSet` instead of a `List`?

seirysjs replied:
changed to `HashSet`

NerijusD commented on Serilog.Sinks.Slack/SlackSink.cs:
```diff
@@ -215,6 +218,20 @@ private void AddAttachmentField(ref Attachment attachment, Field field)
             attachment.Fields.Add(field);
         }
 
+        private string GetPropertyFromException(LogEvent logEvent, OverridableProperties overridableProperty, string defaultValue)
```
Could this be named differently? This method name looks strange. Why it has part `FromException` ?

seirysjs replied:
Renamed to `GetPropertyFromLogEvent`.

NerijusD commented:
@TrapperHell will you be able to give another look at this at some point this week?

seirysjs commented:
@TrapperHell changed version from `2.2.1` to `2.2.2`

TrapperHell commented:
We need your help with this one @mgibas. The NuGet push failed - seems like the API key has expired or is no longer valid.

NerijusD commented:
Hello @TrapperHell, @mgibas,

Any updates on publishing new package?

mgibas commented:
@TrapperHell token regenerated - see ya next year!
